packages/shared-crypto/lib/keyring.ts

```diff
@@ -0,0 +1,314 @@
+      }
+    }
+
+    this.addresses = [...this.addresses, ...activeAddresses]
+
+    return activeAddresses.map(this._getNonSensitiveAddressData)
+  }
+
+  // PRIVATE METHODS
+
+  private _getAddress = (addressHash: AddressHash): SensitiveAddressData => {
+    const address = this.addresses.find(({ hash }) => hash === addressHash)
+
+    if (!address) throw new Error(`Keyring: Could not find address with hash ${addressHash}`)
+
+    return address
+  }
+
+  private _generateAndCacheAddress = (props: GenerateAddressProps): SensitiveAddressData => {
+    const cachedAddressIndex = this.addresses.findIndex(({ index }) => index === props.addressIndex)
+
+    if (cachedAddressIndex >= 0) {
+      if (this.addresses[cachedAddressIndex].privateKey) {
+        return this.addresses[cachedAddressIndex] as SensitiveAddressData
+      } else {
+        this.addresses.splice(cachedAddressIndex, 1)
+      }
+    }
+
+    const address = this._generateAddress(props)
+
+    this.addresses.push(address)
+
+    return address
+  }
+
+  private _generateAddress = ({
+    group,
+    addressIndex = 0,
+    skipAddressIndexes = []
+  }: GenerateAddressProps): SensitiveAddressData => {
+    if (group !== undefined && (!Number.isInteger(group) || group < 0 || group >= TOTAL_NUMBER_OF_GROUPS))
+      throw new Error(`Keyring: Could not generate address in group ${group}, group is invalid`)
+
+    let nextAddressIndex = skipAddressIndexes.includes(addressIndex)
+      ? findNextAvailableAddressIndex(addressIndex, skipAddressIndexes)
+      : addressIndex
+    let newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+
+    while (group !== undefined && addressToGroup(newAddressData.hash, TOTAL_NUMBER_OF_GROUPS) !== group) {
+      nextAddressIndex = findNextAvailableAddressIndex(newAddressData.index, skipAddressIndexes)
+      newAddressData = this._deriveAddressAndKeys(nextAddressIndex)
+    }
+
+    return newAddressData
+  }
+
+  private _deriveAddressesInGroup = (group: number, amount: number, skipIndexes: number[]): SensitiveAddressData[] => {
+    const addresses = []
+    const skipAddressIndexes = Array.from(skipIndexes)
+
+    for (let j = 0; j < amount; j++) {
+      const newAddress = this._generateAddress({ group, skipAddressIndexes })
+      addresses.push(newAddress)
+      skipAddressIndexes.push(newAddress.index)
+    }
+
+    return addresses
+  }
+
+  private _getNonSensitiveAddressData = ({
+    hash,
+    index,
+    publicKey
+  }: SensitiveAddressData): NonSensitiveAddressData => ({ hash, index, publicKey })
+
+  private _initFromMnemonic = (mnemonic: Buffer, passphrase: string) => {
+    if (this.root) throw new Error('Keyring: Secret recovery phrase already provided')
+
+    const isValid = bip39.validateMnemonic(mnemonic.toString())
+    if (!isValid) throw new Error('Keyring: Invalid secret recovery phrase provided')
+
+    const seed = bip39.mnemonicToSeedSync(mnemonic.toString(), passphrase)
+    this.root = bip32.fromSeed(seed)
```

Comment:
This is where we cache the `root` that is used for address generation, just like Metamask: https://github.com/MetaMask/eth-hd-keyring/blob/main/index.js#L300